### PR TITLE
support col match and change to DatasetMatch

### PR DESCRIPTION
**Pull request by VenkataKarthikP**
*Issue, if available: https://github.com/awslabs/deequ/issues/501

Description of changes: 

address review comments from [here](https://github.com/awslabs/deequ/pull/526#pullrequestreview-1810354070) for DatasetMatch.

* Add `matchColumnMappings` for datasetMatch
* Change `DataSynchronization` to `DatasetMatch` 

cc: @rdsharma26
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/scala/com/amazon/deequ/analyzers/DatasetMatchState.scala [hunk 1]
```diff
@@ -17,32 +17,32 @@
...
 /**
- * Represents the state of data synchronization between two DataFrames in Deequ.
- * This state keeps track of the count of synchronized record count and the total record count.
- * It is used to calculate a ratio of synchronization, which is a measure of how well the data
- * in the two DataFrames are synchronized.
+ * Represents the state of datasetMatch between two DataFrames in Deequ.
+ * This state keeps track of the count of matched record count and the total record count.
+ * It is used to calculate a ratio of match, which is a measure of how well the data
+ * in the two DataFrames are matched.
  *
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/analyzers/DatasetMatchState.scala:24]: nit: It measures how well the data in the two DataFrames matches.

File: src/main/scala/com/amazon/deequ/checks/Check.scala [hunk 2]
```diff
@@ -368,37 +368,42 @@ case class Check(
    *   val assertionFunction: Double => Boolean = _ > 0.7
    *
    *   val check = new Check(CheckLevel.Error, "Data Synchronization Check")
-   *     .isDataSynchronized(otherDataFrame, columnMappings, assertionFunction)
+   *     .isDatasetMatched(otherDataFrame, columnMappings, assertionFunction)
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/checks/Check.scala:371]: How about `doesDatasetMatch` ?

File: src/main/scala/com/amazon/deequ/checks/Check.scala [hunk 3]
```diff
@@ -368,37 +368,42 @@ case class Check(
...
-   * @param columnMappings  A map defining the column correlations between the current DataFrame and otherDf.
+   * @param keyColumnMappings  A map defining the column correlations between the current DataFrame and otherDf.
    *                        Keys represent column names in the current DataFrame,
    *                        and values are corresponding column names in otherDf.
    * @param assertion       A function that takes a Double (result of the comparison) and returns a Boolean.
    *                        Defines the condition under which the data in both DataFrames is considered synchronized.
    *                        For example (_ > 0.7) denoting metric value > 0.7 or 70% of records.
+   * @param matchColumnMappings A map defining the column correlations between the current DataFrame and otherDf.
+   *                        These are the columns which we will check for equality, post joining. It's an optional value
+   *                        with defaults to None, which will be derived from `keyColumnMappings` if None.
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/checks/Check.scala:392]: nit: Please align the parameter description lines.

File: src/main/scala/com/amazon/deequ/checks/Check.scala [hunk 4]
```diff
@@ -368,37 +368,42 @@ case class Check(
...
-   *                        of the synchronization check. This object can be used in Deequ's verification suite to
+   *                        of the dataset match check. This object can be used in Deequ's verification suite to
    *                        assert data quality constraints.
    *
    */
-  def isDataSynchronized(otherDf: DataFrame, columnMappings: Map[String, String], assertion: Double => Boolean,
-                         hint: Option[String] = None): Check = {
-    val dataSyncAnalyzer = DataSynchronizationAnalyzer(otherDf, columnMappings, assertion)
-    val constraint = AnalysisBasedConstraint[DataSynchronizationState, Double, Double](dataSyncAnalyzer, assertion,
+  def isDatasetMatched(otherDf: DataFrame,
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/checks/Check.scala:400]: We could change `otherDf` to `otherDataset`. We use `data` as the name of the parameter for the primary dataset in the Verification suite.

**Review by rdsharma26 — COMMENTED** [review]
Left some comments.

File: src/main/scala/com/amazon/deequ/analyzers/DatasetMatchAnalyzer.scala [hunk 5]
```diff
@@ -29,59 +29,67 @@ import scala.util.Try
...
  *                       and its corresponding column in dfToCompare.
+ * @param matchColumnMappings A map defining the column correlations between the current DataFrame and otherDf.
+ *                            These are the columns which we will check for equality, post joining.
+ *                            It's an optional value with defaults to None.
  * @param assertion A function that takes a Double (the match ratio) and returns a Boolean.
  *                  It defines the condition for successful synchronization.
  *
  * Usage:
- * This analyzer is used in Deequ's VerificationSuite based if `isDataSynchronized` check is defined or could be used
+ * This analyzer is used in Deequ's VerificationSuite based if `isDatasetMatched` check is defined or could be used
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/analyzers/DatasetMatchAnalyzer.scala:49]: Can we change `isDatasetMatched` to `doesDatasetMatch` here as well?

File: src/main/scala/com/amazon/deequ/analyzers/DatasetMatchAnalyzer.scala [hunk 6]
```diff
@@ -29,59 +29,67 @@ import scala.util.Try
...
  * manually as well.
  *
  * Example:
- * val analyzer = DataSynchronizationAnalyzer(dfToCompare, Map("col1" -> "col2"), _ > 0.8)
+ * val analyzer = DatasetMatchAnalyzer(dfToCompare, Map("col1" -> "col2"), _ > 0.8)
  * val verificationResult = VerificationSuite().onData(df).addAnalyzer(analyzer).run()
  *
  * // or could do something like below
- * val verificationResult = VerificationSuite().onData(df).isDataSynchronized(dfToCompare, Map("col1" -> "col2"),
+ * val verificationResult = VerificationSuite().onData(df).isDatasetMatched(dfToCompare, Map("col1" -> "col2"),
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/analyzers/DatasetMatchAnalyzer.scala:57]: And same comment here.

**Review by rdsharma26 — APPROVED** [review]
2 minor comments, rest looks good. Thanks again for the PR!

**Comment by VenkataKarthikP** [comment]
> 2 minor comments, rest looks good. Thanks again for the PR!

thanks for review @rdsharma26 ... updated doc strings, could you please re-approve and merge PR. thanks

**Pull request merged**